crates/bevy_picking/src/mesh_picking/ray_cast/intersections.rs

```diff
@@ -66,160 +66,134 @@ pub fn ray_mesh_intersection<I: TryInto<usize> + Clone + Copy>(
         }
 
-        for triangle in indices.chunks_exact(3) {
-            let [a, b, c] = [
-                triangle[0].try_into().ok()?,
-                triangle[1].try_into().ok()?,
-                triangle[2].try_into().ok()?,
-            ];
-
-            let triangle_index = Some(a);
-            let tri_vertex_positions = &[
-                Vec3::from(positions[a]),
-                Vec3::from(positions[b]),
-                Vec3::from(positions[c]),
-            ];
-            let tri_normals = vertex_normals.map(|normals| {
-                [
-                    Vec3::from(normals[a]),
-                    Vec3::from(normals[b]),
-                    Vec3::from(normals[c]),
-                ]
-            });
-
-            let Some(hit) = triangle_intersection(
-                tri_vertex_positions,
-                tri_normals.as_ref(),
-                closest_hit_distance,
-                &mesh_space_ray,
-                backface_culling,
-            ) else {
-                continue;
-            };
-
-            closest_hit = Some(RayMeshHit {
-                point: mesh_transform.transform_point3(hit.point),
-                normal: mesh_transform.transform_vector3(hit.normal),
-                barycentric_coords: hit.barycentric_coords,
-                distance: mesh_transform
-                    .transform_vector3(mesh_space_ray.direction * hit.distance)
-                    .length(),
-                triangle: hit.triangle.map(|tri| {
-                    [
-                        mesh_transform.transform_point3(tri[0]),
-                        mesh_transform.transform_point3(tri[1]),
-                        mesh_transform.transform_point3(tri[2]),
-                    ]
-                }),
-                triangle_index,
-            });
-            closest_hit_distance = hit.distance;
-        }
+        indices
+            .chunks_exact(3)
+            .fold(
+                (f32::MAX, None),
+                |(closest_distance, closest_hit), triangle| {
+                    let [Ok(a), Ok(b), Ok(c)] = [
+                        triangle[0].try_into(),
+                        triangle[1].try_into(),
+                        triangle[2].try_into(),
+                    ] else {
+                        return (closest_distance, closest_hit);
+                    };
+
+                    let tri_vertices = match [positions.get(a), positions.get(b), positions.get(c)]
+                    {
+                        [Some(a), Some(b), Some(c)] => {
+                            [Vec3::from(*a), Vec3::from(*b), Vec3::from(*c)]
+                        }
+                        _ => return (closest_distance, closest_hit),
+                    };
+
+                    match ray_triangle_intersection(&ray, &tri_vertices, backface_culling) {
+                        Some(hit) if hit.distance >= 0. && hit.distance < closest_distance => {
+                            (hit.distance, Some((a, hit)))
```

Comment:
Why are we returning `a` here, isn't that the index of the first triangle vertex?